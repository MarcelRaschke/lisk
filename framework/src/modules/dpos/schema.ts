/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const delegatesUserNamesSchema = {
	$id: '/dpos/userNames',
	type: 'object',
	properties: {
		registeredDelegates: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['username', 'address'],
				properties: {
					username: {
						dataType: 'string',
						fieldNumber: 1,
					},
					address: {
						dataType: 'bytes',
						fieldNumber: 2,
					},
				},
			},
		},
	},
	required: ['registeredDelegates'],
};

export const dposModuleParamsSchema = {
	$id: '/dpos/params',
	type: 'object',
	required: ['activeDelegates', 'standbyDelegates', 'delegateListRoundOffset'],
	additionalProperties: true,
	properties: {
		activeDelegates: {
			dataType: 'uint32',
		},
		standbyDelegates: {
			dataType: 'uint32',
		},
		delegateListRoundOffset: {
			dataType: 'uint32',
		},
	},
};

export const dposAccountSchema = {
	type: 'object',
	properties: {
		delegate: {
			type: 'object',
			fieldNumber: 1,
			properties: {
				username: { dataType: 'string', fieldNumber: 1 },
				pomHeights: {
					type: 'array',
					items: { dataType: 'uint32' },
					fieldNumber: 2,
				},
				consecutiveMissedBlocks: { dataType: 'uint32', fieldNumber: 3 },
				lastForgedHeight: { dataType: 'uint32', fieldNumber: 4 },
				isBanned: { dataType: 'boolean', fieldNumber: 5 },
				totalVotesReceived: { dataType: 'uint64', fieldNumber: 6 },
			},
			required: [
				'username',
				'pomHeights',
				'consecutiveMissedBlocks',
				'lastForgedHeight',
				'isBanned',
				'totalVotesReceived',
			],
		},
		sentVotes: {
			type: 'array',
			fieldNumber: 2,
			items: {
				type: 'object',
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
				required: ['delegateAddress', 'amount'],
			},
		},
		unlocking: {
			type: 'array',
			fieldNumber: 3,
			items: {
				type: 'object',
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
					unvoteHeight: {
						dataType: 'uint32',
						fieldNumber: 3,
					},
				},
				required: ['delegateAddress', 'amount', 'unvoteHeight'],
			},
		},
	},
	default: {
		delegate: {
			username: '',
			pomHeights: [],
			consecutiveMissedBlocks: 0,
			lastForgedHeight: 0,
			isBanned: false,
			totalVotesReceived: BigInt(0),
		},
		sentVotes: [],
		unlocking: [],
	},
};

export const voteWeightsSchema = {
	$id: '/dpos/voteWeights',
	type: 'object',
	properties: {
		voteWeights: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				properties: {
					round: {
						dataType: 'uint32',
						fieldNumber: 1,
					},
					delegates: {
						type: 'array',
						fieldNumber: 2,
						items: {
							type: 'object',
							properties: {
								address: {
									dataType: 'bytes',
									fieldNumber: 1,
								},
								voteWeight: {
									dataType: 'uint64',
									fieldNumber: 2,
								},
							},
						},
					},
				},
			},
		},
	},
	required: ['voteWeights'],
};
