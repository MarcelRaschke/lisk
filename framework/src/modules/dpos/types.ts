/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { BlockHeader } from '@liskhq/lisk-chain';

export interface RegisteredDelegate {
	readonly username: string;
	readonly address: Buffer;
}

export interface RegisteredDelegates {
	registeredDelegates: RegisteredDelegate[];
}

export interface DelegatePersistedUsernames {
	readonly registeredDelegates: RegisteredDelegate[];
}

export interface UnlockingAccountAsset {
	readonly delegateAddress: Buffer;
	readonly amount: bigint;
	readonly unvoteHeight: number;
}

export interface UnlockingInfoJSON {
	readonly delegateAddress: string;
	readonly amount: string;
	readonly unvoteHeight: number;
	readonly minUnlockHeight: number;
}

export interface VoteAccountAsset {
	readonly delegateAddress: Buffer;
	// Amount for some delegate can be updated
	amount: bigint;
}

export interface DPOSAccountProps {
	dpos: {
		delegate: {
			username: string;
			pomHeights: number[];
			consecutiveMissedBlocks: number;
			lastForgedHeight: number;
			isBanned: boolean;
			totalVotesReceived: bigint;
		};
		sentVotes: VoteAccountAsset[];
		unlocking: UnlockingAccountAsset[];
	};
}

export interface UnlockTransactionAssetContext {
	readonly unlockObjects: ReadonlyArray<UnlockingAccountAsset>;
}

export interface RegisterTransactionAssetContext {
	readonly username: string;
}

export interface VoteTransactionAssetContext {
	readonly votes: ReadonlyArray<VoteAccountAsset>;
}

export interface BlockHeaderAssetForDPOS {
	readonly seedReveal: Buffer;
	readonly maxHeightPreviouslyForged: number;
	readonly maxHeightPrevoted: number;
}

export interface PomTransactionAssetContext {
	readonly header1: BlockHeader<BlockHeaderAssetForDPOS>;
	readonly header2: BlockHeader<BlockHeaderAssetForDPOS>;
}

export interface DelegateWeight {
	readonly address: Buffer;
	readonly voteWeight: bigint;
}

export interface VoteWeight {
	readonly round: number;
	readonly delegates: ReadonlyArray<DelegateWeight>;
}

export type VoteWeights = VoteWeight[];

export interface DecodedVoteWeights {
	voteWeights: VoteWeights;
}

type Grow<T, A extends T[]> = ((x: T, ...xs: A) => void) extends (...a: infer X) => void
	? X
	: never;
type GrowToSize<T, A extends T[], N extends number> = {
	readonly 0: A;
	readonly 1: GrowToSize<T, Grow<T, A>, N>;
}[A['length'] extends N ? 0 : 1];

export type FixedLengthArray<T, N extends number> = GrowToSize<T, [], N>;

// Look for a way to define buffer type with fixed size
export type RandomSeed = Buffer;
