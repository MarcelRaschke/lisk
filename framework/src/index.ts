/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export {
	Transaction,
	transactionSchema,
	readGenesisBlockJSON,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	getAccountSchemaWithDefault,
	getGenesisBlockHeaderAssetSchema,
	blockHeaderAssetSchema,
} from '@liskhq/lisk-chain';
export {
	BaseModule,
	BaseAsset,
	TokenModule,
	TokenTransferAsset,
	SequenceModule,
	KeysModule,
	KeysRegisterAsset,
	DPoSModule,
	DPoSRegisterAsset,
	DPoSVoteAsset,
	DPoSUnlockAsset,
	DPoSPoMAsset,
} from './modules';
export { Application } from './application';
export { systemDirs } from './system_dirs';
export { BasePlugin, BlockHeaderJSON, PluginInfo, PluginCodec } from './plugins/base_plugin';
export { IPCChannel } from './controller/channels';
export type { BaseChannel } from './controller/channels';
export type { EventsDefinition, EventCallback } from './controller/event';
export type { ActionsDefinition, ActionHandler } from './controller/action';
export * as testing from './testing';
export * from './types';
