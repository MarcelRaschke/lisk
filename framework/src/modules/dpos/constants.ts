/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

// State store related constants
export const CHAIN_STATE_DELEGATE_USERNAMES = 'dpos:delegateUsernames';
export const CHAIN_STATE_DELEGATE_VOTE_WEIGHTS = 'dpos:delegateVoteWeights';

export const TEN_UNIT = BigInt(10) * BigInt(10) ** BigInt(8);
export const MAX_VOTE = 10;
export const MAX_UNLOCKING = 20;
export const MAX_INT64 = '9223372036854775807';
export const AMOUNT_MULTIPLIER_FOR_VOTES = BigInt(10) * BigInt(10) ** BigInt(8);
export const WAIT_TIME_VOTE = 2000;
export const WAIT_TIME_SELF_VOTE = 260000;
export const VOTER_PUNISH_TIME = 260000;
export const SELF_VOTE_PUNISH_TIME = 780000;
export const MAX_PUNISHABLE_BLOCK_HEIGHT_DIFFERENCE = 260000;
export const MAX_POM_HEIGHTS = 5;
export const MAX_CONSECUTIVE_MISSED_BLOCKS = 50;
export const MAX_LAST_FORGED_HEIGHT_DIFF = 260000;

export const DEFAULT_ACTIVE_DELEGATE = 101;
export const DEFAULT_STANDBY_DELEGATE = 2;
export const DEFAULT_ROUND_OFFSET = 2;

// Vote weight is capped at the self-vote * DEFAULT_VOTE_WEIGHT_CAP_RATE
export const DEFAULT_VOTE_WEIGHT_CAP_RATE = 10;
export const DEFAULT_STANDBY_THRESHOLD = BigInt(1000) * BigInt(10) ** BigInt(8);

// Punishment period is 780k block height by default
export const PUNISHMENT_PERIOD = 780000;
