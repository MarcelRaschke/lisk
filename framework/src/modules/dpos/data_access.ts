/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { StateStore } from '../../types';
import {
	DecodedVoteWeights,
	DelegatePersistedUsernames,
	RegisteredDelegates,
	VoteWeights,
} from './types';
import { CHAIN_STATE_DELEGATE_USERNAMES, CHAIN_STATE_DELEGATE_VOTE_WEIGHTS } from './constants';
import { delegatesUserNamesSchema, voteWeightsSchema } from './schema';

export const getRegisteredDelegates = async (
	store: StateStore,
): Promise<DelegatePersistedUsernames> => {
	const usernamesBuffer = await store.chain.get(CHAIN_STATE_DELEGATE_USERNAMES);
	if (!usernamesBuffer) {
		return { registeredDelegates: [] };
	}
	const parsedUsernames = codec.decode<RegisteredDelegates>(
		delegatesUserNamesSchema,
		usernamesBuffer,
	);

	return parsedUsernames;
};

export const setRegisteredDelegates = async (
	store: StateStore,
	usernames: DelegatePersistedUsernames,
): Promise<void> => {
	usernames.registeredDelegates.sort((a, b) => a.address.compare(b.address));

	await store.chain.set(
		CHAIN_STATE_DELEGATE_USERNAMES,
		codec.encode(delegatesUserNamesSchema, usernames),
	);
};

export const getVoteWeights = async (stateStore: StateStore): Promise<VoteWeights> => {
	const voteWeights = await stateStore.chain.get(CHAIN_STATE_DELEGATE_VOTE_WEIGHTS);
	if (!voteWeights) {
		return [];
	}

	const voteWeightsDecoded = codec.decode<DecodedVoteWeights>(voteWeightsSchema, voteWeights);
	return voteWeightsDecoded.voteWeights;
};

export const setVoteWeights = async (
	stateStore: StateStore,
	voteWeights: VoteWeights,
): Promise<void> => {
	await stateStore.chain.set(
		CHAIN_STATE_DELEGATE_VOTE_WEIGHTS,
		codec.encode(voteWeightsSchema, { voteWeights }),
	);
};

export const deleteVoteWeightsUntilRound = async (
	round: number,
	stateStore: StateStore,
): Promise<void> => {
	const voteWeights = await getVoteWeights(stateStore);
	const newVoteWeights = voteWeights.filter(vw => vw.round >= round);
	await setVoteWeights(stateStore, newVoteWeights);
};
