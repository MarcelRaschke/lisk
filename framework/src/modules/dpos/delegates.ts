/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { getAddressFromPublicKey, hash } from '@liskhq/lisk-cryptography';
import { Account } from '@liskhq/lisk-chain';
import { Logger } from '../../logger/logger';
import { Consensus, StateStore } from '../../types';
import { DelegateWeight, DPOSAccountProps } from './types';
import { getRegisteredDelegates, getVoteWeights, setVoteWeights } from './data_access';
import {
	DEFAULT_ACTIVE_DELEGATE,
	DEFAULT_STANDBY_DELEGATE,
	DEFAULT_STANDBY_THRESHOLD,
	DEFAULT_VOTE_WEIGHT_CAP_RATE,
	MAX_CONSECUTIVE_MISSED_BLOCKS,
	MAX_LAST_FORGED_HEIGHT_DIFF,
} from './constants';
import { isCurrentlyPunished } from './utils';

export const shuffleDelegateList = (
	previousRoundSeed1: Buffer,
	addresses: ReadonlyArray<Buffer>,
): ReadonlyArray<Buffer> => {
	const delegateList = [...addresses].map(delegate => ({
		address: delegate,
	})) as { address: Buffer; roundHash: Buffer }[];

	for (const delegate of delegateList) {
		const seedSource = Buffer.concat([previousRoundSeed1, delegate.address]);
		delegate.roundHash = hash(seedSource);
	}

	delegateList.sort((delegate1, delegate2) => {
		const diff = delegate1.roundHash.compare(delegate2.roundHash);
		if (diff !== 0) {
			return diff;
		}

		return delegate1.address.compare(delegate2.address);
	});

	return delegateList.map(delegate => delegate.address);
};

export const pickStandByDelegate = (
	delegateWeights: ReadonlyArray<DelegateWeight>,
	randomSeed: Buffer,
): number => {
	const seedNumber = randomSeed.readBigUInt64BE();
	const totalVoteWeight = delegateWeights.reduce(
		(prev, current) => prev + BigInt(current.voteWeight),
		BigInt(0),
	);

	let threshold = seedNumber % totalVoteWeight;
	for (let i = 0; i < delegateWeights.length; i += 1) {
		const voteWeight = BigInt(delegateWeights[i].voteWeight);
		if (voteWeight > threshold) {
			return i;
		}
		threshold -= voteWeight;
	}

	return -1;
};

export const updateDelegateList = async ({
	round,
	randomSeeds,
	stateStore,
	activeDelegates,
	standbyDelegates,
	consensus,
}: {
	round: number;
	randomSeeds: ReadonlyArray<Buffer>;
	consensus: Consensus;
	stateStore: StateStore;
	activeDelegates: number;
	standbyDelegates: number;
}): Promise<void> => {
	if (!randomSeeds.length) {
		throw new Error('Random seed must be provided');
	}

	const voteWeights = await getVoteWeights(stateStore);

	const voteWeight = voteWeights.find(vw => vw.round === round);
	if (!voteWeight) {
		throw new Error(`Corresponding vote weight for round ${round.toString()} not found`);
	}
	// Expect that voteWeight is stored in order of voteWeight and address
	const hasStandbySlot = voteWeight.delegates.length > activeDelegates;
	const activeDelegateAddresses = voteWeight.delegates
		.slice(0, activeDelegates)
		.map(vw => vw.address);
	const standbyDelegateAddresses: Buffer[] = [];
	const standbyDelegateVoteWeights = hasStandbySlot
		? voteWeight.delegates.slice(activeDelegates)
		: [];

	// If standby delegates are less or equal to what required
	// Then don't choose based on random seed and consider those as standby
	if (
		standbyDelegateVoteWeights.length > 0 &&
		standbyDelegateVoteWeights.length <= standbyDelegates
	) {
		for (const delegate of standbyDelegateVoteWeights) {
			standbyDelegateAddresses.push(delegate.address);
		}
		// If standby delegates are more than what required then choose based on random seed
	} else if (standbyDelegateVoteWeights.length > standbyDelegates) {
		for (let i = 0; i < standbyDelegates; i += 1) {
			const standbyDelegateIndex = pickStandByDelegate(
				standbyDelegateVoteWeights,
				randomSeeds[i % randomSeeds.length],
			);
			if (standbyDelegateIndex < 0) {
				throw new Error('Fail to pick standby delegate');
			}
			standbyDelegateAddresses.push(standbyDelegateVoteWeights[standbyDelegateIndex].address);
			standbyDelegateVoteWeights.splice(standbyDelegateIndex, 1);
		}
	}

	const delegates = activeDelegateAddresses.concat(standbyDelegateAddresses);
	const shuffledDelegates = shuffleDelegateList(randomSeeds[0], delegates);

	const delegatesList = shuffledDelegates.map(address => ({
		address,
		isConsensusParticipant: !standbyDelegateAddresses.includes(address),
	}));

	await consensus.updateDelegates(delegatesList);
};

export const createVoteWeightsSnapshot = async ({
	height,
	stateStore,
	round,
	logger,
	voteWeightCapRate = DEFAULT_VOTE_WEIGHT_CAP_RATE,
	activeDelegates = DEFAULT_ACTIVE_DELEGATE,
	standbyDelegates = DEFAULT_STANDBY_DELEGATE,
	standbyThreshold = DEFAULT_STANDBY_THRESHOLD,
}: {
	height: number;
	stateStore: StateStore;
	round: number;
	voteWeightCapRate?: number;
	activeDelegates?: number;
	standbyDelegates?: number;
	standbyThreshold?: bigint;
	logger: Logger;
}): Promise<void> => {
	logger.debug(`Creating vote weight snapshot for round: ${round.toString()}`);

	const delegateUserNames = await getRegisteredDelegates(stateStore);

	const delegates: Account<DPOSAccountProps>[] = await Promise.all(
		delegateUserNames.registeredDelegates.map(async delegate =>
			stateStore.account.get<Account<DPOSAccountProps>>(delegate.address),
		),
	);

	// Update totalVotesReceived to voteWeight equivalent before sorting
	for (const account of delegates) {
		// If the account is being punished, then consider them as vote weight 0
		if (isCurrentlyPunished(height, account.dpos.delegate.pomHeights)) {
			account.dpos.delegate.totalVotesReceived = BigInt(0);
			continue;
		}
		const selfVote = account.dpos.sentVotes.find(vote =>
			vote.delegateAddress.equals(account.address),
		);
		const cappedValue = (selfVote?.amount ?? BigInt(0)) * BigInt(voteWeightCapRate);
		if (account.dpos.delegate.totalVotesReceived > cappedValue) {
			account.dpos.delegate.totalVotesReceived = cappedValue;
		}
	}

	delegates.sort((a, b) => {
		const diff = b.dpos.delegate.totalVotesReceived - a.dpos.delegate.totalVotesReceived;
		if (diff > BigInt(0)) {
			return 1;
		}
		if (diff < BigInt(0)) {
			return -1;
		}

		return a.address.compare(b.address);
	});

	const activeDelegatesList = [];
	const standbyDelegatesList = [];
	for (const account of delegates) {
		// If the account is banned, do not include in the list
		if (account.dpos.delegate.isBanned) {
			continue;
		}

		// Select active delegate first
		if (activeDelegatesList.length < activeDelegates) {
			activeDelegatesList.push({
				address: account.address,
				voteWeight: account.dpos.delegate.totalVotesReceived,
			});
			continue;
		}

		// If account has more than threshold, save it as standby
		if (account.dpos.delegate.totalVotesReceived >= standbyThreshold) {
			standbyDelegatesList.push({
				address: account.address,
				voteWeight: account.dpos.delegate.totalVotesReceived,
			});
			continue;
		}

		// From here, it's below threshold
		// Below threshold, but prepared array does not have enough selected delegate
		if (standbyDelegatesList.length < standbyDelegates) {
			// In case there was 1 standby delegate who has more than threshold
			standbyDelegatesList.push({
				address: account.address,
				voteWeight: account.dpos.delegate.totalVotesReceived,
			});
			continue;
		}
		break;
	}

	const delegateVoteWeights = activeDelegatesList.concat(standbyDelegatesList);
	const voteWeight = {
		round,
		delegates: delegateVoteWeights,
	};

	// Save result to the chain state with round number
	const voteWeights = await getVoteWeights(stateStore);
	const voteWeightsIndex = voteWeights.findIndex(vw => vw.round === round);

	if (voteWeightsIndex === -1) {
		voteWeights.push(voteWeight);
	} else {
		voteWeights[voteWeightsIndex] = voteWeight;
	}

	await setVoteWeights(stateStore, voteWeights);
};

export const updateDelegateProductivity = async ({
	height,
	blockTime,
	generatorPublicKey,
	blockTimestamp,
	stateStore,
	consensus,
}: {
	height: number;
	blockTime: number;
	generatorPublicKey: Buffer;
	blockTimestamp: number;
	stateStore: StateStore;
	consensus: Consensus;
}): Promise<void> => {
	const lastBlock = stateStore.chain.lastBlockHeaders[0];
	const expectedForgingAddresses = (await consensus.getDelegates()).map(d => d.address);

	const missedBlocks = Math.ceil((blockTimestamp - lastBlock.timestamp) / blockTime) - 1;
	const forgerAddress = getAddressFromPublicKey(generatorPublicKey);
	const forgerIndex = expectedForgingAddresses.findIndex(address => address.equals(forgerAddress));
	// Update consecutive missed block
	for (let i = 0; i < missedBlocks; i += 1) {
		const rawIndex = (forgerIndex - 1 - i) % expectedForgingAddresses.length;
		const index = rawIndex >= 0 ? rawIndex : rawIndex + expectedForgingAddresses.length;
		const missedForgerAddress = expectedForgingAddresses[index];
		const missedForger = await stateStore.account.get<DPOSAccountProps>(missedForgerAddress);
		missedForger.dpos.delegate.consecutiveMissedBlocks += 1;

		// Ban the missed forger if both consecutive missed block and last forged block diff condition are met
		if (
			missedForger.dpos.delegate.consecutiveMissedBlocks > MAX_CONSECUTIVE_MISSED_BLOCKS &&
			height - missedForger.dpos.delegate.lastForgedHeight > MAX_LAST_FORGED_HEIGHT_DIFF
		) {
			missedForger.dpos.delegate.isBanned = true;
		}
		await stateStore.account.set(missedForgerAddress, missedForger);
	}

	// Reset consecutive missed block
	const forger = await stateStore.account.get<DPOSAccountProps>(forgerAddress);
	forger.dpos.delegate.consecutiveMissedBlocks = 0;
	forger.dpos.delegate.lastForgedHeight = height;
	await stateStore.account.set(forgerAddress, forger);
};
