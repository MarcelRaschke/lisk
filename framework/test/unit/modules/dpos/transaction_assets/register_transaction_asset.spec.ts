/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { validator } from '@liskhq/lisk-validator';
import { RegisterTransactionAsset } from '../../../../../src/modules/dpos/transaction_assets/register_transaction_asset';
import { ApplyAssetContext } from '../../../../../src/types';
import {
	RegisterTransactionAssetContext,
	DPoSModule,
	DPOSAccountProps,
} from '../../../../../src/modules/dpos';
import { CHAIN_STATE_DELEGATE_USERNAMES } from '../../../../../src/modules/dpos/constants';
import * as testing from '../../../../../src/testing';

const { StateStoreMock } = testing.mocks;

describe('RegisterTransactionAsset', () => {
	const lastBlockHeight = 200;
	let transactionAsset: RegisterTransactionAsset;
	let context: ApplyAssetContext<RegisterTransactionAssetContext>;
	let sender: any;
	let stateStoreMock: testing.mocks.StateStoreMock;

	beforeEach(() => {
		sender = testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {});
		stateStoreMock = new StateStoreMock({
			accounts: [sender],
			lastBlockHeaders: [{ height: lastBlockHeight }] as any,
		});
		transactionAsset = new RegisterTransactionAsset();

		const asset = {
			username: 'delegate',
		};

		const transaction = {
			senderAddress: sender.address,
		} as any;

		context = testing.createApplyAssetContext({
			transaction,
			asset,
			stateStore: stateStoreMock,
		});

		jest.spyOn(stateStoreMock.account, 'get');
		jest.spyOn(stateStoreMock.account, 'set');
		jest.spyOn(stateStoreMock.chain, 'get');
		jest.spyOn(stateStoreMock.chain, 'set');
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(0);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('registerDelegate');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('#validate', () => {
		describe('schema validation', () => {
			it('should throw error when username is more than 20 characters', () => {
				const asset: RegisterTransactionAssetContext = { username: 'abchd1563890087376368' };

				const errors = validator.validate(transactionAsset.schema, asset);
				expect(errors).toHaveLength(1);
				expect(errors[0].message).toInclude('must NOT have more than 20 characters');
			});

			it('should throw error when username empty string', () => {
				const asset: RegisterTransactionAssetContext = { username: '' };

				const errors = validator.validate(transactionAsset.schema, asset);
				expect(errors).toHaveLength(1);
				expect(errors[0].message).toInclude('must NOT have fewer than 1 characters');
			});
		});

		it('should not throw error if valid username is provided', () => {
			// Arrange
			const asset: RegisterTransactionAssetContext = { username: 'obelisk' };

			// Act & Assert
			expect(() => transactionAsset.validate({ asset } as any)).not.toThrow(
				'The username is in unsupported format',
			);
		});

		it('should throw error when username includes capital letter', () => {
			// Arrange
			const asset: RegisterTransactionAssetContext = { username: 'Obelisk' };

			// Act & Assert
			expect(() => transactionAsset.validate({ asset } as any)).toThrow(
				'The username is in unsupported format',
			);
		});

		it('should throw error when username includes forbidden character', () => {
			// Arrange
			const asset: RegisterTransactionAssetContext = { username: 'obe^lis' };

			// Act & Assert
			expect(() => transactionAsset.validate({ asset } as any)).toThrowErrorMatchingSnapshot();
		});

		it('should throw error when username includes forbidden null character', () => {
			// Arrange
			const asset: RegisterTransactionAssetContext = { username: 'obe\0lisk' };

			// Act & Assert
			expect(() => transactionAsset.validate({ asset } as any)).toThrow(
				'The username is in unsupported format',
			);
		});

		it('should not throw error when username includes only integers', () => {
			// Arrange
			const asset: RegisterTransactionAssetContext = { username: '123456789987' };

			// Act & Assert
			expect(() => transactionAsset.validate({ asset } as any)).not.toThrow();
		});

		it('should not throw error when username looks like legacy account address', () => {
			// Arrange
			const asset: RegisterTransactionAssetContext = { username: '123456789987l' };

			// Act & Assert
			expect(() => transactionAsset.validate({ asset } as any)).not.toThrow();
		});

		it.each(['!', '@', '$', '&', '_', '.'])(
			'should not throw error when username includes whitelisted special character "%s"',
			(character: string) => {
				// Arrange
				const asset: RegisterTransactionAssetContext = { username: `obe${character}lisk` };

				// Act & Assert
				expect(() => transactionAsset.validate({ asset } as any)).not.toThrow();
			},
		);

		it.each(['%', '#', ')', '(', '-', ' '])(
			'should throw error when username includes non-whitelisted special character "%s"',
			(character: string) => {
				// Arrange
				const asset: RegisterTransactionAssetContext = { username: `obe${character}lisk` };

				// Act & Assert
				expect(() => transactionAsset.validate({ asset } as any)).toThrow(
					'The username is in unsupported format',
				);
			},
		);
	});

	describe('#apply', () => {
		it('should call state store', async () => {
			// Act
			await transactionAsset.apply(context);

			// Assert
			expect(stateStoreMock.account.get).toHaveBeenCalledWith(context.transaction.senderAddress);
			expect(stateStoreMock.account.set).toHaveBeenCalledWith(sender.address, {
				...sender,
				dpos: {
					...sender.dpos,
					delegate: {
						...sender.dpos.delegate,
						username: context.asset.username,
						lastForgedHeight: lastBlockHeight + 1,
					},
				},
			});
		});

		it('should not throw errors', async () => {
			// Act
			await stateStoreMock.account.set(
				sender.address,
				testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {
					address: sender.address,
				}),
			);

			// Act & Assert
			await expect(transactionAsset.apply(context)).resolves.toBeUndefined();
		});

		it('should throw error when username is taken', async () => {
			const delegatesUserNamesSchema = {
				$id: '/dpos/userNames',
				type: 'object',
				properties: {
					registeredDelegates: {
						type: 'array',
						fieldNumber: 1,
						items: {
							type: 'object',
							properties: {
								username: {
									dataType: 'string',
									fieldNumber: 1,
								},
								address: {
									dataType: 'bytes',
									fieldNumber: 2,
								},
							},
						},
					},
				},
				required: ['registeredDelegates'],
			};

			const secondAccount = testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {
				dpos: { delegate: { username: 'myuser' } },
			});
			await stateStoreMock.account.set(secondAccount.address, secondAccount);
			context.asset = { username: 'myuser' };

			await stateStoreMock.chain.set(
				CHAIN_STATE_DELEGATE_USERNAMES,
				codec.encode(delegatesUserNamesSchema, {
					registeredDelegates: [
						{
							username: context.asset.username,
							address: Buffer.from('random'),
						},
					],
				}),
			);

			await expect(transactionAsset.apply(context)).rejects.toThrow(
				`Username ${context.asset.username} is already registered.`,
			);
		});

		it('should throw error when account is already delegate', async () => {
			const defaultVal = testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {});
			await stateStoreMock.account.set(
				sender.address,
				testing.fixtures.createDefaultAccount([DPoSModule], {
					address: sender.address,
					dpos: {
						delegate: {
							...defaultVal.dpos.delegate,
							username: 'alreadydelegate',
						},
					},
				}),
			);

			await expect(transactionAsset.apply(context)).rejects.toThrow(
				'Account is already a delegate',
			);
		});

		it('should set lastForgedHeight to the lastBlock height + 1', async () => {
			// Arrange
			await stateStoreMock.account.set(
				sender.address,
				testing.fixtures.createDefaultAccount([DPoSModule], { address: sender.address }),
			);

			// Act
			await transactionAsset.apply(context);

			// Assert
			const updatedSender = await stateStoreMock.account.get(sender.address);
			expect(updatedSender.dpos.delegate.lastForgedHeight).toEqual(lastBlockHeight + 1);
		});
	});
});
