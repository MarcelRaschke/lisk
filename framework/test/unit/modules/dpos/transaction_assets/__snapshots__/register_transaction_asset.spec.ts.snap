// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterTransactionAsset #validate should throw error when username includes forbidden character 1`] = `"The username is in unsupported format"`;

exports[`RegisterTransactionAsset constructor should have valid schema 1`] = `
Object {
  "$id": "lisk/dpos/register",
  "properties": Object {
    "username": Object {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 20,
      "minLength": 1,
    },
  },
  "required": Array [
    "username",
  ],
  "type": "object",
}
`;
